File: Sean的leetcode/FB/Word Search.py
###
# DFS
# Time Complexity: O(n*m*4^(k)) n:row m:col k:len of word
# Space Complexity: O(k)
###

class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        if not board or not word:
            return False
        for i in xrange(len(board)):
            for j in xrange(len(board[0])):
                if self.dfs(board, i, j, word, 0):
                    return True
        return False
        
    def dfs(self, board, i, j, word, idx):
        if len(word) == idx:
            return True
        if i < 0 or j < 0 or i >= len(board) or j >= len(board[0]) or word[idx] != board[i][j]:
            return False
        board[i][j] = '#'
        exist = self.dfs(board, i+1, j, word, idx+1) or self.dfs(board, i-1, j, word, idx+1) or self.dfs(board, i, j-1, word, idx+1) or self.dfs(board, i, j+1, word, idx+1)
        board[i][j] = word[idx]
        return exist

File: MyLeetCode/Leetcode/Word Search.cpp
//https://leetcode.com/discuss/30514/accepted-c-code-with-dfs
class Solution {
public:
    bool search(vector<vector<char>>& board, string word, int start, int row, int col)
    {
        int max_row = board.size() - 1;
        int max_col = board[0].size() - 1;

        if (start == word.size())
            return true;

        if (row < 0 || row > max_row || col < 0 || col > max_col)
            return false;

        char org_c = board[row][col];
        if (board[row][col] == word[start])
        {
            board[row][col] = 0;
            if (search(board, word, start+1, row, col+1)
                || search(board, word, start+1, row, col-1)
                || search(board, word, start+1, row+1, col)
                || search(board, word, start+1, row-1, col))
                return true;
        }
        board[row][col] = org_c;
        return false;
    }

    bool exist(vector<vector<char>>& board, string word) {

        if (board.size() == 0) return false;

        int r = board.size();
        int c = board[0].size();

        for (int i = 0; i < r; i++)
            for (int j = 0; j < c; j++)
                if (search(board, word, 0, i, j))
                    return true;
        return false;
    }
};